File: 88_3-D_Tic-Tac-Toe/python/qubit.py
#!/usr/bin/python3

# Ported from the BASIC source for 3D Tic Tac Toe
# in BASIC Computer Games, by David H. Ahl
# The code originated from Dartmouth College

from enum import Enum


class Move(Enum):
    """ Game status and types of machine move """
    HUMAN_WIN = 0
    MACHINE_WIN = 1
    DRAW = 2
    MOVES = 3
    LIKES = 4
    TAKES = 5
    GET_OUT = 6
    YOU_FOX = 7
    NICE_TRY = 8
    CONCEDES = 9

class Player(Enum):
    EMPTY = 0
    HUMAN = 1
    MACHINE = 2

class TicTacToe3D:
    """ The game logic for 3D Tic Tac Toe and the machine opponent """

    def __init__(self):
        
        # 4x4x4 board keeps track of which player occupies each place
        # and used by machine to work out its strategy
        self.board = [0] * 64

        # starting move
        self.corners = [0, 48, 51, 3, 12, 60, 63, 15,
                        21, 38, 22, 37, 25, 41, 26, 42]

        # lines to check for end game
        self.lines = [[0, 1, 2, 3],
                      [4, 5, 6, 7],
                      [8, 9, 10, 11],
                      [12, 13, 14, 15],
                      [16, 17, 18, 19],
                      [20, 21, 22, 23],
                      [24, 25, 26, 27],
                      [28, 29, 30, 31],
                      [32, 33, 34, 35],
                      [36, 37, 38, 39],
                      [40, 41, 42, 43],
                      [44, 45, 46, 47],
                      [48, 49, 50, 51],
                      [52, 53, 54, 55],
                      [56, 57, 58, 59],
                      [60, 61, 62, 63],
                      [0, 16, 32, 48],
                      [4, 20, 36, 52],
                      [8, 24, 40, 56],
                      [12, 28, 44, 60],
                      [1, 17, 33, 49],
                      [5, 21, 37, 53],
                      [9, 25, 41, 57],
                      [13, 29, 45, 61],
                      [2, 18, 34, 50],
                      [6, 22, 38, 54],
                      [10, 26, 42, 58],
                      [14, 30, 46, 62],
                      [3, 19, 35, 51],
                      [7, 23, 39, 55],
                      [11, 27, 43, 59],
                      [15, 31, 47, 63],
                      [0, 4, 8, 12],
                      [16, 20, 24, 28],
                      [32, 36, 40, 44],
                      [48, 52, 56, 60],
                      [1, 5, 9, 13],
                      [17, 21, 25, 29],
                      [33, 37, 41, 45],
                      [49, 53, 57, 61],
                      [2, 6, 10, 14],
                      [18, 22, 26, 30],
                      [34, 38, 42, 46],
                      [50, 54, 58, 62],
                      [3, 7, 11, 15],
                      [19, 23, 27, 31],
                      [35, 39, 43, 47],
                      [51, 55, 59, 63],
                      [0, 5, 10, 15],
                      [16, 21, 26, 31],
                      [32, 37, 42, 47],
                      [48, 53, 58, 63],
                      [12, 9, 6, 3],
                      [28, 25, 22, 19],
                      [44, 41, 38, 35],
                      [60, 57, 54, 51],
                      [0, 20, 40, 60],
                      [1, 21, 41, 61],
                      [2, 22, 42, 62],
                      [3, 23, 43, 63],
                      [48, 36, 24, 12],
                      [49, 37, 25, 13],
                      [50, 38, 26, 14],
                      [51, 39, 27, 15],
                      [0, 17, 34, 51],
                      [4, 21, 38, 55],
                      [8, 25, 42, 59],
                      [12, 29, 46, 63],
                      [48, 33, 18, 3],
                      [52, 37, 22, 7],
                      [56, 41, 26, 11],
                      [60, 45, 30, 15],
                      [0, 21, 42, 63],
                      [15, 26, 37, 48],
                      [3, 22, 41, 60],
                      [12, 25, 38, 51]]

    def get(self, x, y, z):
        m = self.board[4*(4*z + y) + x]
        if m == 40:
            return Player.MACHINE
        elif m == 8:
            return Player.HUMAN
        else:
            return Player.EMPTY

    def move3D(self, x, y, z, player):
        m = 4*(4*z + y) + x
        return self.move(m, player)

    def move(self, m, player):
        if self.board[m] > 1:
            return False
        
        if player == Player.MACHINE:
            self.board[m] = 40
        else:
            self.board[m] = 8
        return True

    def get3DPosition(self, m):
        x = m % 4
        y = (m // 4) % 4
        z = m // 16
        return x, y, z

    def evaluateLines(self):
        self.lineValues = [0] * 76
        for j in range(76):
            value = 0
            for k in range(4):
                value += self.board[self.lines[j][k]]
            self.lineValues[j] = value

    def strategyMarkLine(self, i):
        for j in range(4):
            m = self.lines[i][j]
            if self.board[m] == 0:
                self.board[m] = 1

    def clearStrategyMarks(self):
        for i in range(64):
            if self.board[i] == 1:
                self.board[i] = 0

    def markAndMove(self, vlow, vhigh, vmove):
        """ 
        mark lines that can potentially win the game for the human
        or the machine and choose best place to play
        """
        for i in range(76):
            value = 0
            for j in range(4):
                value += self.board[self.lines[i][j]]
            self.lineValues[i] = value
            if vlow <= value < vhigh:
                if value > vlow:
                    return self.moveTriple(i)
                self.strategyMarkLine(i)
        self.evaluateLines()

        for i in range(76):
            value = self.lineValues[i]
            if value == 4 or value == vmove:
                return self.moveDiagonals(i, 1)
        return None

    def machineMove(self):
        """ machine works out what move to play """
        self.clearStrategyMarks()

        self.evaluateLines()
        for value, event in [(32, self.humanWin),
                             (120, self.machineWin),
                             (24, self.blockHumanWin)]:
            for i in range(76):
                if self.lineValues[i] == value:
                    return event(i)

        m = self.markAndMove(80, 88, 43)
        if m != None:
            return m

        self.clearStrategyMarks()

        m = self.markAndMove(16, 24, 11)
        if m != None:
            return m

        for k in range(18):
            value = 0
            for i in range(4*k, 4*k+4):
                for j in range(4):
                    value += self.board[self.lines[i][j]]
            if (32 <= value < 40) or (72 <= value < 80):
                for s in [1, 0]:
                    for i in range(4*k, 4*k+4):
                        m = self.moveDiagonals(i, s)
                        if m != None:
                            return m

        self.clearStrategyMarks()

        for y in self.corners:
            if self.board[y] == 0:
                return (Move.MOVES, y)

        for i in range(64):
            if self.board[i] == 0:
                return (Move.LIKES, i)

        return (Move.DRAW, -1)
            
    def humanWin(self, i):
        return (Move.HUMAN_WIN, -1, i)

    def machineWin(self, i):
        for j in range(4):
            m = self.lines[i][j]
            if self.board[m] == 0:
                return (Move.MACHINE_WIN, m, i)
        return None

    def blockHumanWin(self, i):
        for j in range(4):
            m = self.lines[i][j]
            if self.board[m] == 0:
                return (Move.NICE_TRY, m)
        return None


    def moveTriple(self, i):
        """ make two lines-of-3 or prevent human from doing this """
        for j in range(4):
            m = self.lines[i][j]
            if self.board[m] == 1:
                if self.lineValues[i] < 40:
                    return (Move.YOU_FOX, m)
                else:
                    return (Move.GET_OUT, m)
        return (Move.CONCEDES, -1)

    # choose move in corners or center boxes of square 4x4
    def moveDiagonals(self, i, s):
        if 0 < (i % 4) < 3:
            jrange = [1, 2]
        else:
            jrange = [0, 3]
        for j in jrange:
            m = self.lines[i][j]
            if self.board[m] == s:
                return (Move.TAKES, m)
        return None

class Qubit:
    
    def moveCode(self, board, m):
        x, y, z = board.get3DPosition(m)
        return "{:d}{:d}{:d}".format(z+1, y+1, x+1)
        
    def showWin(self, board, i):
        for m in board.lines[i]:
            print(self.moveCode(board, m))

    def showBoard(self, board):
        c = " YM"
        for z in range(4):
            for y in range(4):
                print("   " * y, end="")
                for x in range(4):
                    p = board.get(x, y, z)
                    print("({})      ".format(c[p.value]), end="")
                print("\n")
            print("\n")
        
    def humanMove(self, board):
        print("")
        c = "1234"
        while True:
            h = input("Your move?\n")
            if h == "1":
                return False
            if h == "0":
                self.showBoard(board)
                continue
            if (len(h) == 3) and (h[0] in c) and (h[1] in c) and (h[2] in c):
                x = c.find(h[2])
                y = c.find(h[1])
                z = c.find(h[0])
                if board.move3D(x, y, z, Player.HUMAN):
                    break
                
                print("That square is used. Try again.")
            else:
                print("Incorrect move. Retype it--")

        return True
        
    def play(self):
        print("Qubic\n")
        print("Create Computing Morristown, New Jersey\n\n\n")
        while True:
            c = input("Do you want instructions?\n")
            if len(c) >= 1 and (c[0] in "ynYN"):
                break
            print("Incorrect answer. Please type 'yes' or 'no.")
            
        c = c.lower()
        if c[0] == 'y':
            print("The game is Tic-Tac-Toe in a 4 x 4 x 4 cube.")
            print("Each move is indicated by a 3 digit number, with each")
            print("digit between 1 and 4 inclusive.  The digits indicate the")
            print("level, row, and column, respectively, of the occupied")
            print("place.\n")

            print("To print the playing board, type 0 (zero) as your move.")
            print("The program will print the board with your moves indicated")
            print("with a (Y), the machine's moves with an (M), and")
            print("unused squares with a ( ).\n")

            print("To stop the program run, type 1 as your move.\n\n")

        
            
        play_again = True
        while play_again:
            board = TicTacToe3D()

            while True:
                s = input("Do you want to move first?\n")
                if len(s) >= 1 and (s[0] in "ynYN"):
                    break
                print("Incorrect answer. Please type 'yes' or 'no'.")

            skipHuman = s[0] in "nN"

            move_text = ["Machine moves to",
                         "Machine likes",
                         "Machine takes",
                         "Let's see you get out of this:  Machine moves to",
                         "You fox.  Just in the nick of time, machine moves to",
                         "Nice try. Machine moves to"]

            while True:
                if not skipHuman:
                    if not self.humanMove(board):
                        break
                skipHuman = False

                m = board.machineMove()
                if m[0] == Move.HUMAN_WIN:
                    print("You win as follows,")
                    self.showWin(board, m[2])
                    break
                elif m[0] == Move.MACHINE_WIN:
                    print("Machine moves to {}, and wins as follows".format(self.moveCode(board, m[1])))
                    self.showWin(board, m[2])
                    break
                elif m[0] == Move.DRAW:
                    print("The game is a draw.")
                    break
                elif m[0] == Move.CONCEDES:
                    print("Machine concedes this game.")
                    break
                else:
                    print(move_text[m[0].value - Move.MOVES.value])
                    print(self.moveCode(board, m[1]))
                    board.move(m[1], Player.MACHINE)

                self.showBoard(board)

            print(" ")
            while True:
                x = input("Do you want to try another game\n")
                if len(x) >= 1 and x[0] in "ynYN":
                    break
                print("Incorrect answer. Please Type 'yes' or 'no'.")

            play_again = x[0] in "yY"
        
        

if __name__ == "__main__":
    game = Qubit()
    game.play()

File: 00_Alternate_Languages/88_3-D_Tic-Tac-Toe/javascript/qubit.js
// QUBIT
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;

    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");

                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var xa = [];
var la = [];
var ma = [[],
          [,1,2,3,4],    // 1
          [,5,6,7,8],    // 2
          [,9,10,11,12], // 3
          [,13,14,15,16],    // 4
          [,17,18,19,20],    // 5
          [,21,22,23,24],    // 6
          [,25,26,27,28],    // 7
          [,29,30,31,32],    // 8
          [,33,34,35,36],    // 9
          [,37,38,39,40],    // 10
          [,41,42,43,44],    // 11
          [,45,46,47,48],    // 12
          [,49,50,51,52],    // 13
          [,53,54,55,56],    // 14
          [,57,58,59,60],    // 15
          [,61,62,63,64],    // 16
          [,1,17,33,49], // 17
          [,5,21,37,53],    // 18
          [,9,25,41,57],   // 19
          [,13,29,45,61], // 20
          [,2,18,34,50], // 21
          [,6,22,38,54],    // 22
          [,10,26,42,58],  // 23
          [,14,30,46,62],   // 24
          [,3,19,35,51], // 25
          [,7,23,39,55],    // 26
          [,11,27,43,59],  // 27
          [,15,31,47,63], // 28
          [,4,20,36,52], // 29
          [,8,24,40,56], // 30
          [,12,28,44,60],    // 31
          [,16,32,48,64],    // 32
          [,1,5,9,13],   // 33
          [,17,21,25,29],    // 34
          [,33,37,41,45],    // 35
          [,49,53,57,61],    // 36
          [,2,6,10,14],  // 37
          [,18,22,26,30],    // 38
          [,34,38,42,46],    // 39
          [,50,54,58,62],    // 40
          [,3,7,11,15],  // 41
          [,19,23,27,31],    // 42
          [,35,39,43,47],    // 43
          [,51,55,59,63],    // 44
          [,4,8,12,16],  // 45
          [,20,24,28,32],    // 46
          [,36,40,44,48],    // 47
          [,52,56,60,64],    // 48
          [,1,6,11,16],  // 49
          [,17,22,27,32],    // 50
          [,33,38,43,48],    // 51
          [,49,54,59,64],    // 52
          [,13,10,7,4],  // 53
          [,29,26,23,20],    // 54
          [,45,42,39,36],    // 55
          [,61,58,55,52],    // 56
          [,1,21,41,61], // 57
          [,2,22,42,62], // 58
          [,3,23,43,63], // 59
          [,4,24,44,64], // 60
          [,49,37,25,13],    // 61
          [,50,38,26,14],    // 62
          [,51,39,27,15],    // 63
          [,52,40,28,16],    // 64
          [,1,18,35,52], // 65
          [,5,22,39,56], // 66
          [,9,26,43,60], // 67
          [,13,30,47,64],    // 68
          [,49,34,19,4], // 69
          [,53,38,23,8], // 70
          [,57,42,27,12],    // 71
          [,61,46,31,16],    // 72
          [,1,22,43,64], // 73
          [,16,27,38,49],    // 74
          [,4,23,42,61], // 75
          [,13,26,39,52] // 76
          ];
var ya = [,1,49,52,4,13,61,64,16,22,39,23,38,26,42,27,43];

function show_board()
{
    for (xx = 1; xx <= 9; xx++)
        print("\n");
    for (i = 1; i <= 4; i++) {
        for (j = 1; j <= 4; j++) {
            str = "";
            for (i1 = 1; i1 <= j; i1++)
                str += "   ";
            for (k = 1; k <= 4; k++) {
                q = 16 * i + 4 * j + k - 20;
                if (xa[q] == 0)
                    str += "( )      ";
                if (xa[q] == 5)
                    str += "(M)      ";
                if (xa[q] == 1)
                    str += "(Y)      ";
                if (xa[q] == 1 / 8)
                    str += "( )      ";
            }
            print(str + "\n");
            print("\n");
        }
        print("\n");
        print("\n");
    }
}

function process_board()
{
    for (i = 1; i <= 64; i++) {
        if (xa[i] == 1 / 8)
            xa[i] = 0;
    }
}

function check_for_lines()
{
    for (s = 1; s <= 76; s++) {
        j1 = ma[s][1];
        j2 = ma[s][2];
        j3 = ma[s][3];
        j4 = ma[s][4];
        la[s] = xa[j1] + xa[j2] + xa[j3] + xa[j4];
    }
}

function show_square(m)
{
    k1 = Math.floor((m - 1) / 16) + 1;
    j2 = m - 16 * (k1 - 1);
    k2 = Math.floor((j2 - 1) / 4) + 1;
    k3 = m - (k1 - 1) * 16 - (k2 - 1) * 4;
    m = k1 * 100 + k2 * 10 + k3;
    print(" " + m + " ");
}

function select_move() {
    if (i % 4 <= 1) {
        a = 1;
    } else {
        a = 2;
    }
    for (j = a; j <= 5 - a; j += 5 - 2 * a) {
        if (xa[ma[i][j]] == s)
            break;
    }
    if (j > 5 - a)
        return false;
    xa[ma[i][j]] = s;
    m = ma[i][j];
    print("MACHINE TAKES");
    show_square(m);
    return true;
}

// Main control section
async function main()
{
    print(tab(33) + "QUBIC\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    while (1) {
        print("DO YOU WANT INSTRUCTIONS");
        str = await input();
        str = str.substr(0, 1);
        if (str == "Y" || str == "N")
            break;
        print("INCORRECT ANSWER.  PLEASE TYPE 'YES' OR 'NO'");
    }
    if (str == "Y") {
        print("\n");
        print("THE GAME IS TIC-TAC-TOE IN A 4 X 4 X 4 CUBE.\n");
        print("EACH MOVE IS INDICATED BY A 3 DIGIT NUMBER, WITH EACH\n");
        print("DIGIT BETWEEN 1 AND 4 INCLUSIVE.  THE DIGITS INDICATE THE\n");
        print("LEVEL, ROW, AND COLUMN, RESPECTIVELY, OF THE OCCUPIED\n");
        print("PLACE.  \n");
        print("\n");
        print("TO PRINT THE PLAYING BOARD, TYPE 0 (ZERO) AS YOUR MOVE.\n");
        print("THE PROGRAM WILL PRINT THE BOARD WITH YOUR MOVES INDI-\n");
        print("CATED WITH A (Y), THE MACHINE'S MOVES WITH AN (M), AND\n");
        print("UNUSED SQUARES WITH A ( ).  OUTPUT IS ON PAPER.\n");
        print("\n");
        print("TO STOP THE PROGRAM RUN, TYPE 1 AS YOUR MOVE.\n");
        print("\n");
        print("\n");
    }
    while (1) {
        for (i = 1; i <= 64; i++)
            xa[i] = 0;
        z = 1;
        print("DO YOU WANT TO MOVE FIRST");
        while (1) {
            str = await input();
            str = str.substr(0, 1);
            if (str == "Y" || str == "N")
                break;
            print("INCORRECT ANSWER.  PLEASE TYPE 'YES' OR 'NO'");
        }
        while (1) {
            while (1) {
                print(" \n");
                print("YOUR MOVE");
                j1 = parseInt(await input());
                if (j1 == 0) {
                    show_board();
                    continue;
                }
                if (j1 == 1)
                    return;
                k1 = Math.floor(j1 / 100);
                j2 = j1 - k1 * 100;
                k2 = Math.floor(j2 / 10);
                k3 = j2 - k2 * 10;
                m = 16 * k1 + 4 * k2 + k3 - 20;
                if (k1 < 1 || k2 < 1 || k3 < 1 || k1 > 4 || k2 > 4 || k3 >> 4) {
                    print("INCORRECT MOVE, RETYPE IT--");
                } else {
                    process_board();
                    if (xa[m] != 0) {
                        print("THAT SQUARE IS USED, TRY AGAIN.\n");
                    } else {
                        break;
                    }
                }
            }
            xa[m] = 1;
            check_for_lines();
            status = 0;
            for (j = 1; j <= 3; j++) {
                for (i = 1; i <= 76; i++) {
                    if (j == 1) {
                        if (la[i] != 4)
                            continue;
                        print("YOU WIN AS FOLLOWS");
                        for (j = 1; j <= 4; j++) {
                            m = ma[i][j];
                            show_square(m);
                        }
                        status = 1;
                        break;
                    }
                    if (j == 2) {
                        if (la[i] != 15)
                            continue;
                        for (j = 1; j <= 4; j++) {
                            m = ma[i][j];
                            if (xa[m] != 0)
                                continue;
                            xa[m] = 5;
                            print("MACHINE MOVES TO ");
                            show_square(m);
                        }
                        print(", AND WINS AS FOLLOWS");
                        for (j = 1; j <= 4; j++) {
                            m = ma[i][j];
                            show_square(m);
                        }
                        status = 1;
                        break;
                    }
                    if (j == 3) {
                        if (la[i] != 3)
                            continue;
                        print("NICE TRY, MACHINE MOVES TO");
                        for (j = 1; j <= 4; j++) {
                            m = ma[i][j];
                            if (xa[m] != 0)
                                continue;
                            xa[m] = 5;
                            show_square(m);
                            status = 2;
                        }
                        break;
                    }
                }
                if (i <= 76)
                    break;
            }
            if (status == 2)
                continue;
            if (status == 1)
                break;
            // x = x; non-useful in original
            i = 1;
            do {
                la[i] = xa[ma[i][1]] + xa[ma[i][2]] + xa[ma[i][3]] + xa[ma[i][4]];
                l = la[i];
                if (l == 10) {
                    for (j = 1; j <= 4; j++) {
                        if (xa[ma[i][j]] == 0)
                            xa[ma[i][j]] = 1 / 8;
                    }
                }
            } while (++i <= 76) ;
            check_for_lines();
            i = 1;
            do {
                if (la[i] == 0.5) {
                    s = 1 / 8;
                    select_move();
                    break;
                }
                if (la[i] == 5 + 3 / 8) {
                    s = 1 / 8;
                    select_move();
                    break;
                }
            } while (++i <= 76) ;
            if (i <= 76)
                continue;

            process_board();

            i = 1;
            do {
                la[i] = xa[ma[i][1]] + xa[ma[i][2]] + xa[ma[i][3]] + xa[ma[i][4]];
                l = la[i];
                if (l == 2) {
                    for (j = 1; j <= 4; j++) {
                        if (xa[ma[i][j]] == 0)
                            xa[ma[i][j]] = 1 / 8;
                    }
                }
            } while (++i <= 76) ;
            check_for_lines();
            i = 1;
            do {
                if (la[i] == 0.5) {
                    s = 1 / 8;
                    select_move();
                    break;
                }
                if (la[i] == 1 + 3 / 8) {
                    s = 1 / 8;
                    select_move();
                    break;
                }
            } while (++i <= 76) ;
            if (i <= 76)
                continue;

            for (k = 1; k <= 18; k++) {
                p = 0;
                for (i = 4 * k - 3; i <= 4 * k; i++) {
                    for (j = 1; j <= 4; j++)
                        p += xa[ma[i][j]];
                }
                if (p == 4 || p == 9) {
                    s = 1 / 8;
                    for (i = 4 * k - 3; i <= 4 * k; i++) {
                        if (select_move())
                            break;
                    }
                    s = 0;
                }
            }
            if (k <= 18)
                continue
            process_board();
            z = 1;
            do {
                if (xa[ya[z]] == 0)
                    break;
            } while (++z < 17) ;
            if (z >= 17) {
                for (i = 1; i <= 64; i++) {
                    if (xa[i] == 0) {
                        xa[i] = 5;
                        m = i;
                        print("MACHINE LIKES");
                        break;
                    }
                }
                if (i > 64) {
                    print("THE GAME IS A DRAW.\n");
                    break;
                }
            } else {
                m = ya[z];
                xa[m] = 5;
                print("MACHINE MOVES TO");
            }
            show_square(m);
        }
        print(" \n");
        print("DO YOU WANT TO TRY ANOTHER GAME");
        while (1) {
            str = await input();
            str = str.substr(0, 1);
            if (str == "Y" || str == "N")
                break;
            print("INCORRECT ANSWER. PLEASE TYPE 'YES' OR 'NO'");
        }
        if (str == "N")
            break;
    }
}

main();
